### ISSUE TITLE
About the wall-clock running time

### ISSUE THREAD

**Issue by zyushun** [nil0x9/flash-muon#4]
Hi! Thanks for the great work! Really enjoy reading your code.

I am testing the running time for the mmt_kernel. I observe a significantly slower runing time than the default implementation, along with a non-zero error. Am I testing it correctly?  Need some help here. Tested on single RTX4090

```
XX^T: flash-muon Time taken: 3.1212656497955322 seconds
XX^T: default Time taken: 0.07193422317504883 seconds
XX^T: flash-muon error torch.norm(fast_muon_result - default_result): 4.302063465118408
```

Thanks in advance!

Here is my simple  test script.

```
import torch
from matmul_transpose_triton import matmul_transpose
import time

X = torch.randn(256, 256, dtype=torch.float32, device="cuda")

start = time.time()
fast_muon_result = matmul_transpose(X)
end = time.time()
print(f"XX^T: flash-muon Time taken: {end - start} seconds")



start = time.time()
default_result = X@X.T
end = time.time()
print(f"XX^T: default Time taken: {end - start} seconds")


error = torch.norm(fast_muon_result - default_result)
print(f"XX^T: flash-muon error torch.norm(fast_muon_result - default_result): {error}")

```

**Comment by nil0x9** [comment]
Hi, Yushun. Thanks for trying out the repo! There are two issues concerning this test:
1. Our codebase generally doesn't do well for small-dim matrices like `X = torch.randn(256, 256, dtype=torch.float32, device="cuda")`. We recommend using this codebase for larger dimensional cases. You can see in the README file that for 4090 the method is on-par with torch native matmul for 1024-dim and surpasses baselines for dim >= 2048.
2. GPU ops are generally asynchronous. So I think using `time.time()` would give you wrong results generally. A more standard practice would be to use cuda Event to measure wall clock time. We give a script in [benchmark.py](https://github.com/nil0x9/flash-muon/blob/main/benchmark.py) in the main branch. Feel free to try that out (you can comment out the NS part if you wish only to test the `matmul_transpose` func)!

The following is the results I got on RTX 4090 (freshly tested a minute ago):
```sh
python benchmark.py 

benchmark matmul transponse:
                    device   dim  flash(ms)  torch(ms)  compiled(ms)
0  NVIDIA GeForce RTX 4090  1024      0.041      0.018         0.025
1  NVIDIA GeForce RTX 4090  2048      0.072      0.102         0.102
2  NVIDIA GeForce RTX 4090  4096      0.483      0.828         0.831
3  NVIDIA GeForce RTX 4090  8192      3.374      6.874         6.901

benchmark zeropower:
                    device   dim  flash(ms)  torch(ms)  compiled(ms)
0  NVIDIA GeForce RTX 4090  1024      0.578      0.327         0.348
1  NVIDIA GeForce RTX 4090  2048      1.401      1.704         1.739
2  NVIDIA GeForce RTX 4090  4096     11.097     15.316        14.879
3  NVIDIA GeForce RTX 4090  8192     77.080    112.263       111.611
```

**Comment by nil0x9** [comment]
And do note that the above result is tested under `bfloat16` precision, as consistent with common practices in Muon implementation. I think the code in the main branch can be run in `fp32` too but I never tested that though. The cuda branch only supports half precision inputs.

**Comment by zyushun** [comment]
@nil0x9 Thanks for the quick response. I can reproduce your results now. Interestingly, I observe more acceleration in float32 than in bfloat16. 

```
benchmark matmul transponse:

Performance Results: (dtype = torch.float32)
                    device   dim  flash(ms)  torch(ms)  compiled(ms)
0  NVIDIA GeForce RTX 4090   256      0.071      0.018         0.056
1  NVIDIA GeForce RTX 4090   512      0.071      0.014         0.083
2  NVIDIA GeForce RTX 4090  1024      0.073      0.057         0.088
3  NVIDIA GeForce RTX 4090  2048      0.184      0.441         0.456
4  NVIDIA GeForce RTX 4090  4096      1.172      3.693         3.665
5  NVIDIA GeForce RTX 4090  8192      7.690     27.496        27.554

Error Summary: (dtype = torch.float32)
           Metric  Value
0  mean_abs_error  0.058
1  mean_rel_error  0.004
2  rms_norm_error  0.102
```


But there are still some small errors compared to the default PyTorch implementation (not just for bf16, but also for fp32). What is the possible cause?


```
def calculate_matrix_difference(mat1, mat2):
    """Calculate the difference between two matrices."""
    # Calculate absolute difference
    abs_diff = torch.abs(mat1 - mat2)
    # Calculate relative difference
    rel_diff = abs_diff / (torch.abs(mat1) + 1e-7)
    return {
        'mean_abs_error': abs_diff.mean().item(),
        'mean_rel_error': rel_diff.mean().item(),
        'rms_norm_error': (torch.norm(mat1 - mat2).item()) / math.sqrt(mat1.size(0) * mat1.size(1))
    }
```

**Comment by nil0x9** [comment]
Regarding the error that you mentioned, it's important to note that addition in float numbers is **not commutative**, which means that different algos with different compute orders would result in different outputs. For inputs like `randn(256, 256)` or even larger reduction dimension, the diagonal numbers would be large, thus their absolute error would be large as well (as they fall into area of lower precision ). You can easily verify the position where the largest error occur with `torch.nonzero((result - ref).abs() == (result - ref).abs().max())`. It would most definitely output a diagonal position. 

The problem would quickly vanish when the input is smaller (e.g., scaled by some factor `> 1`). And I think in optimizers this behavior would be tolerated bc: (1) NS is itself not a precise procedure; (2) the input gradient momentum is typically in smaller range as the training continues.

That being said, I think the problem still remains why the error induced by `fp32` inputs is larger than `fp16` counterparts. I think it might have something to do with triton but I'm not sure...

**Comment by zyushun** [comment]
Thanks @nil0x9 for the detailed response! I checked the largest error position and they indeed lie in the diagonal entries. Good to know!

Thanks for the help!

```
Largest error position: tensor([[67, 67]], device='cuda:0')
Largest error position: tensor([[166, 166]], device='cuda:0')
Largest error position: tensor([[245, 245]], device='cuda:0')
Largest error position: tensor([[2002, 2002]], device='cuda:0')
Largest error position: tensor([[2806, 2806]], device='cuda:0')
```
